### PR TITLE
Oauth

### PR DESCRIPTION
**Pull request by Zwork101**
Let's finally get this out of the way!

### CODE REVIEW

File: pysite/database.py [hunk 1]
```diff
@@ -103,9 +103,11 @@ def create_table(self, table_name: str, primary_key: str="id", durability: str="
...
             return True
 
-    def delete(self, table_name: str, primary_key: Optional[str] = None,
-               durability: str = "hard", return_changes: Union[bool, str] = False
-               ) -> Union[Dict[str, Any], None]:
+    def delete(self,
+               table_name: str,
+               primary_key: Union[str, None] = None,
+               durability: str="hard",
+               return_changes: Union[bool, str] = False) -> dict:
```
Comment (gdude2002) [inline pysite/database.py:110]: rude. :P

File: pysite/oauth.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+from flask_dance.contrib.discord import discord
+import requests
+
+from pysite.constants import DISCORD_API_ENDPOINT, SERVER_ID
+from pysite.mixins import DBMixin
+
+_MAIN_BACK = None  # type: pysite.database.RethinkDB
+
+
+class OauthBackend(BaseBackend, DBMixin):
```
Comment (gdude2002) [inline pysite/oauth.py]: DBMixin was never designed for this. It's a mixin for route views, not just any object.

I suggest the following:
* Remove `_MAIN_BACK` from the module. Globals are bad.
* Store the database manager from the route manager on the object, via `self`.
* Move your outer functions into the class.
* Make sure the OauthBackend is accessible to the route manager - it needs to be stored within it.
* Finally, make the OauthBackend available to routes, if any of your routes need it.

File: pysite/database.py (hunk 1 above)
Comment (Zwork101) [reply pysite/database.py:110]: Apeture suggested it last PR :p

File: pysite/views/main/oauth_resp.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    name = "oauth_resp"
+
+    def get(self):
+        code = r.args.get('code')  # Get the code parameter from the URL req
+        if code is None:  # If the code was not supplied, return with a (400)
+            return "Failed Request", 400  # Create error pages
+
+        data = {
+            'client_id': "{client_id}",  # Fill this in
+            'client_secret': "{client_secret}",  # Fill this in
```
Comment (gdude2002) [inline pysite/views/main/oauth_resp.py]: `# fill this in`

File: pysite/views/tests/index.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 from pysite.constants import ValidationTypes
 from pysite.decorators import api_params
 
+
```
Comment (gdude2002) [inline pysite/views/tests/index.py:10]: :eyes:

**Review by gdude2002 — DISMISSED** [review]
Not bad. A few comments there for you, and also:

* We should write a middleware to inject logged-in user info into the request object
* Your PR doesn't lint, and it seems that test coverage has been lowered

File: app_test.py [hunk 5]
```diff
@@ -29,11 +30,15 @@ def create_app(self):
         return app
 
 
+class BaseEndpoints(SiteTest):
```
Comment (gdude2002) [inline app_test.py]: Empty testcase?

File: pysite/mixins.py [hunk 6]
```diff
@@ -59,3 +59,25 @@ def table(self) -> Table:
     @property
     def db(self) -> RethinkDB:
         return self._db()
+
+
+class OauthMixin:
+
+    @classmethod
+    def setup(cls: "DBMixin", manager: "pysite.route_manager.RouteManager", blueprint: Blueprint):
```
Comment (gdude2002) [inline pysite/mixins.py]: Wrong `cls` type hint

File: pysite/mixins.py [hunk 7]
```diff
@@ -59,3 +59,25 @@ def table(self) -> Table:
...
+
+        if hasattr(super(), "setup"):
+            super().setup(manager, blueprint)  # pragma: no cover
+
+        cls._oauth = ref(manager.oauth_backend)
```
Comment (gdude2002) [inline pysite/mixins.py]: I still think we might want to be able to access this class directly, so a property like the following might be nice:

```py
@property
def oauth(self):
    return self._oauth()
```

File: pysite/oauth.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        self.db = manager.db
+        self.key = manager.app.secret_key
+
+    def get(self, *args, **kwargs):  # Not used
+        pass
+
+    def set(self, blueprint, token):
+
+        user = self.get_user()
+        self.join_discord(token["access_token"], user["id"])
```
Comment (gdude2002) [inline pysite/oauth.py:40]: Is there any way we can detect that we needed to do this and let the user know on the site?

File: pysite/oauth.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        sess_id = str(uuid5(uuid4(), self.key))
+        session["session_id"] = sess_id
+
+        self.db.insert("oauth_data", {"id": sess_id,
```
Comment (gdude2002) [inline pysite/oauth.py]: I don't think you're creating this table now.

File: pysite/oauth.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        try:
+            resp = requests.put(DISCORD_API_ENDPOINT + f"guilds/{SERVER_ID}/members/{snowflake}",
+                                data={"access_token": token})  # Have user join our server
+            if resp.status_code != 201:
+                logging.warning(f"Unable to add user ({snowflake}) to server, {resp.json()}")
+        except JSONDecodeError:
+            pass  # User already in server.
+
+    def user_data(self):
+        id = session.get("session_id")
```
Comment (gdude2002) [inline pysite/oauth.py]: Shadowing a builtin

File: pysite/views/main/index.py [hunk 11]
```diff
@@ -1,10 +1,12 @@
 # coding=utf-8
 from pysite.base_route import RouteView
+from pysite.constants import DISCORD_OAUTH_REDIRECT
+from pysite.mixins import OauthMixin
 
 
-class IndexView(RouteView):
+class IndexView(OauthMixin, RouteView):
```
Comment (gdude2002) [inline pysite/views/main/index.py]: No, mixins come after base classes

File: pysite/views/main/index.py [hunk 12]
```diff
@@ -1,10 +1,12 @@
...
     path = "/"
     name = "index"
 
     def get(self):
-        return self.render("main/index.html")
+        return self.render("main/index.html", logged_in=self.logged_in, login_url=DISCORD_OAUTH_REDIRECT)
```
Comment (gdude2002) [inline pysite/views/main/index.py]: `logged_in` is required on all pages. This parameter should be added in the `render()` method of the base class instead. I've already got one somewhere.

Comment (gdude2002) [reply pysite/views/main/index.py]: Also, we should have a login url ourselves that redirects to the Discord OAuth page.

**Review by gdude2002 — CHANGES_REQUESTED** [review]

File: pysite/base_route.py [hunk 13]
```diff
@@ -62,12 +64,12 @@ def setup(cls: "RouteView", manager: "pysite.route_manager.RouteManager", bluepr
...
 
-        if hasattr(super(), "setup"):
-            super().setup(manager, blueprint)
-
         if not cls.path or not cls.name:
             raise RuntimeError("Route views must have both `path` and `name` defined")
 
+        if hasattr(super(), "setup"):
+            super().setup(manager, blueprint)
+
```
Comment (gdude2002) [inline pysite/base_route.py]: Now that the test is fixed, is it necessary to have moved this?

File: pysite/mixins.py [hunk 14]
```diff
@@ -59,3 +59,26 @@ def table(self) -> Table:
     @property
     def db(self) -> RethinkDB:
         return self._db()
+
+
+class OauthMixin:
```
Comment (gdude2002) [inline pysite/mixins.py:64]: This class needs documenting

File: pysite/route_manager.py [hunk 15]
```diff
@@ -27,10 +32,25 @@ def __init__(self):
         self.db = RethinkDB()
         self.log = logging.getLogger()
         self.app.secret_key = os.environ.get("WEBPAGE_SECRET_KEY", "super_secret")
-        self.app.config["SERVER_NAME"] = os.environ.get("SERVER_NAME", "pythondiscord.local:8080")
+        # self.app.config["SERVER_NAME"] = os.environ.get("SERVER_NAME", "pythondiscord.local:8080")
```
Comment (gdude2002) [inline pysite/route_manager.py]: Why?

**Review by gdude2002 — CHANGES_REQUESTED** [review]
I don't see anything about cookies here. Did you decide to remove the automatic server join after all?

File: pysite/base_route.py (hunk 13 above)
Comment (gdude2002) [reply pysite/base_route.py]: Bump

File: pysite/mixins.py (hunk 14 above)
Comment (gdude2002) [reply pysite/mixins.py:64]: Bump

**Review by gdude2002 — CHANGES_REQUESTED** [review]

**Review by gdude2002 — DISMISSED** [review]

**Review by gdude2002 — DISMISSED** [review]

File: app_test.py [hunk 16]
```diff
@@ -79,23 +80,38 @@ def test_invite(self):
         self.assertEqual(response.status_code, 302)
 
     def test_ws_test(self):
-        """ check ws_test responds """
+        """ Check ws_test responds """
         response = self.client.get('/ws_test')
         self.assertEqual(response.status_code, 200)
 
+    def test_oauth_authorized(self):
+        """check oauth redirects """
```
Comment (jerbob) [inline app_test.py]: These docstrings aren't consistent with the others: `""" Check oauth redirects """`

File: app_test.py [hunk 17]
```diff
@@ -380,3 +440,42 @@ def testEcho(self):
         ew.on_open()
         ew.on_message('message')
         ew.on_close()
+
+
+class TestOauthBackend(SiteTest):
+    """ Test cases for the oauth.py file"""
```
Comment (jerbob) [inline app_test.py]: These docstrings aren't consistent with the previous, add a space at the end pls

**Review by jerbob — CHANGES_REQUESTED** [review]

**Review by gdude2002 — APPROVED** [review]

**Review by jerbob — APPROVED** [review]

**Pull request merged**